core/src/com/unciv/logic/civilization/diplomacy/DiplomacyManager.kt

```diff
@@ -588,46 +577,71 @@ class DiplomacyManager() : IsPartOfGameInfoSerialization {
-                        addModifier(DiplomaticModifiers.FulfilledPromiseToNotSettleCitiesNearUs, 10f)
-                    }
-                    DiplomacyFlags.RecentlyAttacked.name -> {
-                        civInfo.cityStateFunctions.askForUnitGifts(otherCiv())
-                    }
-                    // These modifiers don't tick down normally, instead there is a threshold number of turns
-                    DiplomacyFlags.RememberDestroyedProtectedMinor.name -> {    // 125
-                        removeModifier(DiplomaticModifiers.DestroyedProtectedMinor)
-                    }
-                    DiplomacyFlags.RememberAttackedProtectedMinor.name -> {     // 75
-                        removeModifier(DiplomaticModifiers.AttackedProtectedMinor)
-                    }
-                    DiplomacyFlags.RememberBulliedProtectedMinor.name -> {      // 75
-                        removeModifier(DiplomaticModifiers.BulliedProtectedMinor)
-                    }
-                    DiplomacyFlags.RememberSidedWithProtectedMinor.name -> {      // 25
-                        removeModifier(DiplomaticModifiers.SidedWithProtectedMinor)
-                    }
+                if (manageExpiredFlag(flag)) continue@loop
+                flagsCountdown.remove(flag)
+            }
+        }
+    }
+
+    private fun manageCityStateProvideMilitaryUnitFlag(flag: String) {
+        if (flag == DiplomacyFlags.ProvideMilitaryUnit.name && civInfo.isMajorCiv() && otherCiv().isCityState() &&
+            civInfo.gameInfo.civilizations.any { civInfo.isAtWarWith(it) && otherCiv().isAtWarWith(it) }
+        ) {
+            for (unique in civInfo.getMatchingUniques(UniqueType.CityStateMoreGiftedUnits)) {
+                flagsCountdown[DiplomacyFlags.ProvideMilitaryUnit.name] =
+                    flagsCountdown[DiplomacyFlags.ProvideMilitaryUnit.name]!! - unique.params[0].toInt() + 1
+                if (flagsCountdown[DiplomacyFlags.ProvideMilitaryUnit.name]!! <= 0) {
+                    flagsCountdown[DiplomacyFlags.ProvideMilitaryUnit.name] = 0
+                    break
                 }
+            }
+        }
+    }
 
-                flagsCountdown.remove(flag)
+    private fun manageExpiredFlag(flag: String): Boolean {
+        when (flag) {
+            DiplomacyFlags.ResearchAgreement.name -> {
+                if (!otherCivDiplomacy().hasFlag(DiplomacyFlags.ResearchAgreement))
+                    scienceFromResearchAgreement()
+            }
+
+            DiplomacyFlags.DefensivePact.name -> {
+                diplomaticStatus = DiplomaticStatus.Peace
+            }
+            // This is confusingly named - in fact, the civ that has the flag set is the MAJOR civ
+            DiplomacyFlags.ProvideMilitaryUnit.name -> {
+                // Do not unset the flag - they may return soon, and we'll continue from that point on
+                if (civInfo.cities.isEmpty() || otherCiv().cities.isEmpty())
+                    return true
```

Comment:
The name of this function didn't indicate what the return value is, so add a docstring on the function

Reply:
Actually, this check should be *ouside* this function.
This function should be "what do we do when flag expires", the question of "should this flag expire" is a different question which sucks not be dealt with in this function